SNAP-1656 Security support in snappydata

ashetkar opened the pull request:
## Changes proposed in this pull request
Security support in SnappyData.

## Patch testing
precheckin in-progress

## ReleaseNotes changes
NA

## Other PRs 
https://github.com/SnappyDataInc/snappydata/pull/731

ashetkar committed:
SNAP-1770: Changes to get credentials into LeadNodeExecutorContext (#224)

* Set the user credentials in LeadNodeExecutorContext and LeadNodeSmartConnectorOpContext from the LanguageConnectionContext.

* Temporarily fixing failing tests

* * Reverting changes in 6f3da98789695ce3a74c3ced13368d7d5023e6ae which passed credentials.
* Instead, using setting those in LanguageConnectionContext and accessing them from it.
* The credentials are now set into LeadNodeExecutorContext as well as in
  LeadNodeSmartConnectorOpContext.

* * Incorporate review comments: rename password fields to 'authToken'

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1656

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
Changing privilege of CREATE_ALL_BUCKETS from system level to user level (#240)

* Changing privilege of CREATE_ALL_BUCKETS from system level to user level

Commit:
Merge branch 'snappy/master' into SNAP-1656

ashetkar committed:
SNAP-1816 Security in Smart Connector mode (#244)

* Changes to make authentication and authorization work in smart connector mode.
* Users can provide the credentials as confs 'spark.snappydata.store.user' and 'spark.snappydata.store.password'
* Changing privilege of CREATE_ALL_BUCKETS from system level to user level
* Hive metastore tables are READ-ONLY to all.
* Granting privilege on a column table now grants the same privilege
  on its catched batch table to the user.

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1656

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1656

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
Merge branch 'snappy/master' into SNAP-1656

dshirish reviewed (commented):
LGTM

kneeraj commented on gemfire-shared/src/main/java/com/gemstone/gemfire/internal/shared/ClientSharedUtils.java:
```diff
@@ -1287,8 +1287,34 @@ public static RuntimeException newRuntimeException(String message,
   }
 
+  // Convert log4j.Level to java.util.logging.Level
+  public static Level converToJavaLogLevel(org.apache.log4j.Level log4jLevel) {
```
Log changes in this PR? Why was this needed?

ashetkar replied:
This was done to suppress the initial warning messages seen on console at smart connector side, when the hive client gets initialized (and it tries to modify some tables in hive metastore).

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/iapi/sql/dictionary/StatementTablePermission.java:
```diff
@@ -134,7 +135,8 @@ public void check( LanguageConnectionContext lcc,
 	
-		if( ! hasPermissionOnTable( dd, authorizationId, forGrant))
+		if( ! hasPermissionOnTable( dd, authorizationId, forGrant) && !isSelectOnHiveMetastore
+				(getTableDescriptor(dd)))
```
Correct the indentation please.

ashetkar replied:
I went with the existing the indentation of these files. If we follow the indentation policy of spaces here, I think, it may not look better.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/iapi/sql/dictionary/StatementTablePermission.java:
```diff
@@ -146,6 +148,11 @@ public void check( LanguageConnectionContext lcc,
 	} // end of check
 
+	protected boolean isSelectOnHiveMetastore(TableDescriptor td) throws StandardException {
+		return td.getSchemaName().equalsIgnoreCase(Misc.SNAPPY_HIVE_METASTORE) &&
```
Same as above.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/impl/sql/conn/GenericLanguageConnectionContext.java:
```diff
@@ -1425,6 +1426,16 @@ public int getUniqueSavepointID()
 		return nextSavepointId-1;
 	}
 
+	public String getUserName()
```
Indentation and brace location on the same line.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/impl/sql/conn/GenericLanguageConnectionContext.java:
```diff
@@ -1425,6 +1426,16 @@ public int getUniqueSavepointID()
+		return this.userName;
+	}
+
+	public String getAuthToken()
```
Same as above.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/impl/sql/execute/TablePrivilegeInfo.java:
```diff
@@ -222,20 +227,48 @@ public void executeGrantRevoke( Activation activation,
+		doExecuteGrantRevoke(activation, grant, grantees, columnBitSets, actionAllowed, true, td);
+		GemFireStore ms = Misc.getMemStore();
+		// TODO ashetkar Get a better way to check if security is enabled.
+		if (ms.isSnappyStore() && "LDAP".equalsIgnoreCase(ms.getBootProperty(Attribute.AUTH_PROVIDER))
```
A method in Misc has been added above?

vibhaska replied:
@kneeraj Didn't got your comment. 

However I was thinking that above method throws an exception. Is is fine or a similar no throw method be implemented in Misc, and used here checking for null.

ashetkar replied:
What Neeraj was talking about was in another PR, which is now merged into this branch. Will use it here.

kneeraj commented:
LGTM. Minor comments.

kneeraj reviewed (commented):
Minor comments.

ashetkar committed:
[SNAP-1841] Methods to check auth-provider value and get the authservice. (#252)

Added methods to check auth-provider value and get the authservice.

Commit:
Merge branch 'snappy/master' into SNAP-1656

Commit:
* Incorporate review comments:
  Update Misc.isSecurityEnabled() and use it in TablePrivilegeInfo.
* Update sanctionedDataSerializables.txt for LeadNodeExecutionContext.scala
  and LeadNodeSmartConnectorOpContext.scala

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-1656

Commit:
* Remove an obsolete comment.
